arbnode/forwarder.go

```diff
@@ -186,3 +205,162 @@ func (f *TxDropper) StopAndWait() {}
 func (f *TxDropper) Started() bool {
 	return true
 }
+
+type RedisTxForwarder struct {
+	stopwaiter.StopWaiterSafe
+
+	config *ForwarderConfig
+	errors int
+
+	mtx              sync.RWMutex
+	forwarder        *TxForwarder
+	redisCoordinator *redisutil.RedisCoordinator
+	currentTarget    string
+	fallbackTarget   string
+}
+
+func NewRedisTxForwarder(fallbackTarget string, config *ForwarderConfig) *RedisTxForwarder {
+	return &RedisTxForwarder{
+		config:         config,
+		fallbackTarget: fallbackTarget,
+	}
+}
+
+func (f *RedisTxForwarder) PublishTransaction(ctx context.Context, tx *types.Transaction) error {
+	forwarder := f.getForwarder()
+	if forwarder == nil {
+		return ErrNoSequencer
+	}
+	return forwarder.PublishTransaction(ctx, tx)
+}
+
+func (f *RedisTxForwarder) CheckHealth(ctx context.Context) error {
+	forwarder := f.getForwarder()
+	if forwarder == nil {
+		return ErrNoSequencer
+	}
+	return forwarder.CheckHealth(ctx)
+}
+
+// not thread safe vs update and itself
+func (f *RedisTxForwarder) Initialize(ctx context.Context) error {
+	var err error
+	f.redisCoordinator, err = redisutil.NewRedisCoordinator(f.config.RedisUrl)
+	if err != nil {
+		return errors.Wrap(err, "unable to create redis coordinator")
+	}
+	return nil
+}
+
+func (f *RedisTxForwarder) retryAfterError() time.Duration {
+	f.errors++
+	retryIn := f.config.RetryInterval * time.Duration(f.errors)
+	if retryIn > f.config.UpdateInterval {
+		retryIn = f.config.UpdateInterval
+	}
+	return retryIn
+}
+
+// returns true when retry interval is saturated and there is a fallback url available
+func (f *RedisTxForwarder) shouldFallbackToStatic() bool {
+	return f.config.RetryInterval*time.Duration(f.errors+1) >= f.config.UpdateInterval &&
+		f.fallbackTarget != "" && f.fallbackTarget != f.currentTarget
```

Comment:
I think we should always fall back here if `f.currentTarget == ""` which means we don't currently have a forwarding target